### PR TITLE
Port StrictMath.floor and ceil from fdlibm

### PR DESCRIPTION
**Pull request by tanishiking**
- https://www.netlib.org/fdlibm/s_ceil.c
- https://www.netlib.org/fdlibm/s_floor.c

And add corner case tests.

### CODE REVIEW

File: javalib/src/main/scala/java/lang/StrictMath.scala [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+ * is preserved.
+ * ====================================================
+ */
+
+package java.lang
+
+object StrictMath {
+
+  // Ported from https://www.netlib.org/fdlibm/s_floor.c
+  def floor(x: scala.Double): scala.Double = {
```
Comment (sjrd) [inline javalib/src/main/scala/java/lang/StrictMath.scala]: `floor` and `ceil` are among the subset of methods for which `Math.floor` is exactly specified. Therefore, `StrictMath` and `Math` do the same thing. I suggest we keep this implementation inside `Math` for those methods, in order to reduce code size.

File: javalib/src/main/scala/java/lang/StrictMath.scala [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    var highWord = (doubleBits >>> 32).toInt  // sign(1) + exponent(11) + mantissa_high(20)
+    var lowWord = doubleBits.toInt            // mantissa_low(32)
+    val exponent = ((highWord >> 20) & 0x7ff) - 1023
+
+    @inline def reconstructDouble(highWord: Int, lowWord: Int): scala.Double = {
+      val bits = (highWord.toLong << 32) | (lowWord.toLong & 0xffffffffL)
+      Double.longBitsToDouble(bits)
+    }
+
+    if (exponent < 20) { // Case 1: decimal point is withing the high word.
```
Comment (sjrd) [inline javalib/src/main/scala/java/lang/StrictMath.scala]: TBH, this seems like an antiquated way of writing this algorithm. fdlibm was written in a time where 64-bit integers were not native, and it shows. Working with the full `Long` value would be a lot easier to understand, IMO. (not to mention efficient, but that's not important for those slow-path methods)

There is value in directly following fdlibm to guarantee the exact same results. But for an exactly-specified method like `floor`, that argument is not so important. A modern implementation that is more understandable is probably wiser.

File: javalib/src/main/scala/java/lang/StrictMath.scala [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+ * ====================================================
+ */
+
+package java.lang
+
+object StrictMath {
+
+  // Ported from https://www.netlib.org/fdlibm/s_floor.c
+  def floor(x: scala.Double): scala.Double = {
+    val doubleBits = Double.doubleToLongBits(x)
```
Comment (sjrd) [inline javalib/src/main/scala/java/lang/StrictMath.scala]:
```suggestion
    val doubleBits = Double.doubleToRawLongBits(x)
```
No need to normalize NaNs. By default we use the raw variant when it makes no difference.

File: javalib/src/main/scala/java/lang/StrictMath.scala [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        x
+      } else {
+        // if (huge + x > 0.0) {
+        val (adjustedHighWord, adjustedLowWord) =
+          if (highWord < 0) {
+            if (exponent == 20) {
+              (highWord + 1, lowWord)
+            } else {
+              val newLowWord = lowWord + (1 << (52 - exponent)) // + LSB of integer part
+              if (Integer.compareUnsigned(newLowWord, lowWord) < 0) { // carry
```
Comment (sjrd) [inline javalib/src/main/scala/java/lang/StrictMath.scala]:
```suggestion
              if (Integer.unsigned_<(newLowWord, lowWord)) { // carry
```
(but you won't need that with a `Long`-based implementation at all)

File: javalib/src/main/scala/java/lang/StrictMath.scala [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+            (highWord, lowWord)
+          }
+        // clear fractional bits
+        reconstructDouble(adjustedHighWord, adjustedLowWord & ~fractionalMask)
+      }
+    }
+  }
+
+  // Ported from https://www.netlib.org/fdlibm/s_ceil.c
+  def ceil(x: scala.Double): scala.Double = {
```
Comment (sjrd) [inline javalib/src/main/scala/java/lang/StrictMath.scala]: Since these methods are slow paths anyway (only used when the optimizer is turned off), consider delegating to `floor` instead of implementing things twice. The spec is such that `ceil(x)` is the same as `-floor(-x)`.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/StrictMathTest.scala [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    // basic
+    assertExactEquals(5.0, StrictMath.floor(5.0))
+    assertExactEquals(5.0, StrictMath.floor(5.7))
+    assertExactEquals(-6.0, StrictMath.floor(-5.7))
+    assertExactEquals(0.0, StrictMath.floor(0.0))
+    assertExactEquals(-0.0, StrictMath.floor(-0.0))
+    assertExactEquals(0.0, StrictMath.floor(0.5))
+    assertExactEquals(-1.0, StrictMath.floor(-0.5))
+
+    // Special values
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/StrictMathTest.scala]: For consistency with other tests, put those first, and also include `+0.0` and `-0.0`.

File: javalib/src/main/scala/java/lang/StrictMath.scala (hunk 2 above)
Comment (tanishiking) [reply javalib/src/main/scala/java/lang/StrictMath.scala]: Ah true, I tried to port the implementation as is, but yeah splitting to low and high seems only for using 32-bit integer operations for performance (at the time when 64-bit integers weren't native).
Just using 64-bit integer makes it much simpler :+1:

File: javalib/src/main/scala/java/lang/StrictMath.scala (hunk 4 above)
Comment (tanishiking) [reply javalib/src/main/scala/java/lang/StrictMath.scala]: Yup, we no longer need carry with a Long-based implementation.

File: javalib/src/main/scala/java/lang/StrictMath.scala (hunk 1 above)
Comment (tanishiking) [reply javalib/src/main/scala/java/lang/StrictMath.scala]: Ah I see, indeed `StrictMath.floor` and `Math.floor` have exactly the same description.
Only the methods like trigonometric/exponents/logarithmic whose output is strictly defined in `StrictMath` should be in that package :+1:

File: javalib/src/main/scala/java/lang/Math.scala [hunk 7]
```diff
@@ -126,8 +138,52 @@ object Math {
...
+  @inline def ceil(a: scala.Double): scala.Double =
+    linkTimeIf(LinkingInfo.targetPureWasm) {
+      -floor(-a)
+    } {
+      js.Math.ceil(a)
+    }
+  @inline def floor(a: scala.Double): scala.Double = {
+    linkTimeIf(LinkingInfo.targetPureWasm) {
+      // Ported from https://www.netlib.org/fdlibm/s_floor.c
+      var bits = Double.doubleToRawLongBits(a)
```
Comment (sjrd) [inline javalib/src/main/scala/java/lang/Math.scala]: For consistency (and arguably readability), put the pure-Wasm implementation in a separate `private def`?

File: javalib/src/main/scala/java/lang/Math.scala [hunk 8]
```diff
@@ -126,8 +138,52 @@ object Math {
...
+      val exponent = (((bits >>> 52).toInt) & 0x7ff) - 1023
+      if (exponent < 0) { // |x| < 1
+        /* Note: In the original C implementation, code block below was surrounded by
+         * "if (huge + x > 0.0)" (huge = 1.0e300),
+         * which was used to raise the IEEE 754 inexact flag.
+         * In Java, there is no standard way to access these flags,
+         * so this check has no practical effect(?) and commented out.
```
Comment (sjrd) [inline javalib/src/main/scala/java/lang/Math.scala]:
```suggestion
         * so this check has no practical effect and was removed.
```
?

**Review by sjrd — APPROVED** [review]
